protocol/op-contracts-manager-single-release-redesign.md

```diff
@@ -0,0 +1,83 @@
+
+## Authentication (Optional)
+
+Authentication was not originally implemented in OPCM because it solely supported deployments, with the rationale being that there was no immediate need to restrict access to the deploy function. However, with the introduction of upgrade functionality, it’s essential to establish a robust authentication mechanism within OPCM. We now need to determine appropriate access controls and define roles for authorized users who will be permitted to perform upgrades. We can choose to implement this now to prepare for future use in OPCM upgrade features, or we may decide to postpone it until a later stage.
+
+# Alternatives Considered
+
+Continue with the approach of adding additional conditional logic for every new chain version that is released. As noted, this leads to a system that is complex and hard to maintain.
+
+# Proposed Solution
+
+## OPCM Architecture Changes
+
+Having one OPCM per *L1 smart contract release version*** means that we can remove the proxy pattern that currently exists with OPCM. The high level architecture changes can be seen below:
+
+### Previous OPCM Deploys
+Before OPCM was proxied so that it was upgradable. e.g. [`0x18CeC91779995AD14c880e4095456B9147160790`](https://etherscan.io/address/0x18CeC91779995AD14c880e4095456B9147160790)
+```mermaid
+flowchart LR
+    A[Any Account] -->|call| C[Proxy]
+    C -->|delegatecall| D[OPCM]
+```
+
+### New OPCM Deploys
+Now OPCM is not proxied and therefore not upgradable. This is by design though as we want to release a new OPCM for each new L1 smart contract release.
+```mermaid
+flowchart LR
+    A[Any Account] -->|call| B[OPCM - op-contracts/v1.3.0]
+    C[Any Account] -->|call| D[OPCM - op-contracts/v1.6.0]
+```
+
+## Versioning
+
+Now that each OPCM deploy is going to be tethered directly to an L1 smart contract release version, for improved UX, we should provide public getters to expose two pieces of information: 
+
+1. The **current L1 smart contracts** that are deployed via this version of OPCM e.g. `op-contracts/v1.6.0`. This means we will no longer need to pass the standard versions toml [file](https://github.com/ethereum-optimism/optimism/blob/4c015e3a36f8910e2cf8b447d62ab4c44b944cca/packages/contracts-bedrock/scripts/deploy/DeployImplementations.s.sol#L61) around like [this](https://github.com/ethereum-optimism/optimism/blob/4c015e3a36f8910e2cf8b447d62ab4c44b944cca/packages/contracts-bedrock/scripts/deploy/DeployImplementations.s.sol#L1060). The implementation addresses can be added to the OPCM contract as as immutable values. 
+2. The **address of the previous OPCM** contract. 
```

Comment:
Great idea, this makes sense to me

Reply:
Love this, but how would we handle the case where we have to patch the old release? Would we patch the upstream OPCMs as well? Maybe it doesn't matter, but thought I'd call out this edge case nonetheless.

Reply:
Perhaps the patch just points to the prior release and indicates that it replaces a previous version, for example if the `prior` var points to the preceding release and `replaces` points to what it replaces:
- Start with: OPCM 1.7.0 with `prior = address(1.6.0)` and `replaces = null`
- Deploy OPCM 1.7.1 with `prior = address(1.6.0)` and `replaces = address(1.7.0)`

Reply:
One of the reasons I was suggesting to include the address of the previous OPCM contract was so that you technically only ever needed to store the latest address. In other words, you can always find the previous OPCM addresses by traversing the list by continually reading `.prior()`. 

I'm not quite following how we can keep this property if we want to patch an old version of an OPCM contract. For example, let's say we have 3 contracts. `1`, `2` and `3`. 

1. Contract `1.prior()` returns `null`.
2. Contract `2.prior()` returns `1`.
3. Contract `3.prior()` returns `2`. 

If we patch contract `2` such that we now have two contracts `2` and `2-patch`. We will want a call to `3.prior()` to return `2-patch`.
How do we do this? My immediate thought is to have an authenticated setter method on OPCM that allows us to `updatePrior(address patchedVersion)`. Or keep the proxy and swap the implementation address to not have to change any `prior()` values. 

@mds1 - We could implement your suggestion in the `2-patch` contract but how would `3` know to reference the new `2-patch` instead of the old `2` contract?

Reply:
On second thought, the scenario I described might not really happen in reality. 

I don't think it makes sense to make a patch on contract `2` if contract `3` exists.

Reply:
That is good callout to make sure we're aligned, but yes I agree with your conclusion that if release `3` is out, it means `2` is deprecated and we wouldn't patch it

Reply:
I've made your suggestion clear here in this commit: https://github.com/ethereum-optimism/design-docs/pull/156/commits/2011fb5d2c4fa482ecee2127d59376465129142d#diff-a9741395d54e86787fd591639a1e294e53c8ee37ddbbf26f8fb03a596e302b46R53

Reply:
Unresolving this thread, we no longer want to keep a reference to the old contract. 

Patching is useful and having a reference to the previous opcm complicates things.

Reply:
addressed in this commit: https://github.com/ethereum-optimism/design-docs/pull/156/commits/5ee9be2e4d2301af2358a061883bc7f2ce84085f